### ISSUE TITLE
DodgeSizeModifier updating size on panning

### ISSUE THREAD

**Issue by mstainer14** [entronad/graphic#310]
Hi @entronad,

I have a question about the `DodgeSizeModifier`, wheen zooming in we can see that the widths of the bars are being expanded accordingly. However, when panning to left side the sizes of the bars are increasing and when panning to the right the sizes are decreasing.

Is it possible to achieve this behaviour only when scaling the graph? I have not been able to achieve it yet so thought I would post here to obtain some help.

Video included for further explanation:
[Screencast from 09-08-24 17:33:36.webm](https://github.com/user-attachments/assets/5bcb5371-25a0-48f2-896a-8ada2160cab2)

**Comment by entronad** [comment]
Of course, I can help with that. I'm the author of the Graphic library, and I'd be happy to explain what's happening and how you can achieve the desired behavior.

It seems you've encountered an interesting interaction with the `DodgeSizeModifier`. Let's dive into why this is happening and how to fix it.

### The Principle Behind `DodgeSizeModifier`

The `DodgeSizeModifier` is designed to prevent visual conflicts between shapes in the same group by adjusting their sizes. When you use it with bar charts, it calculates the available width for each category on the axis and then divides that space among the bars within that category.

The core of the issue you're observing lies in how this "available width" is determined when you pan the chart. When you pan, the visible range of the coordinate system changes. The `DodgeSizeModifier` recalculates the bar widths based on this new visible range. As you pan to the right, more of the axis comes into view on the left, and the modifier might interpret this as more "space" being available overall, causing the bars to resize. The resizing is happening because the modifier is considering the entire logical range of the axis that is currently "in view," even if some of that is off-screen.

### How to Isolate Resizing to Scaling Only

What you want is for the bar widths to be determined by the **zoom level (scale)** but to remain constant during **panning**. To achieve this, we need to ensure that the calculation for the bar width doesn't change when the chart is panned.

The key is to control the `ratio` parameter of the `Size` aesthetic. Instead of letting the `DodgeSizeModifier` dynamically calculate this based on the visible range, we can set a size that is responsive to the zoom level but not the pan position.

Here is the approach I recommend:

1.  **Remove the `DodgeSizeModifier`:** Since its dynamic recalculation during panning is the source of the issue, we will handle the sizing differently.
2.  **Use `Gesture` and `Coord` state to control size:** We can listen to zoom gestures and update the size of the bars based on the current scale of the coordinate system. The `CoordState` provides information about the current visible range and scale.

Here is a code example demonstrating how you can implement this. I'll use a `StatefulWidget` to manage the bar width dynamically based on the zoom level.

```dart
import 'package:flutter/material.dart';
import 'package:graphic/graphic.dart';

class FixedPanBarChart extends StatefulWidget {
  const FixedPanBarChart({super.key});

  @override
  State<FixedPanBarChart> createState() => _FixedPanBarChartState();
}

class _FixedPanBarChartState extends State<FixedPanBarChart> {
  final GlobalKey<CoordState> _coordKey = GlobalKey();
  // Initial size ratio for the bars.
  double _barSizeRatio = 0.5;

  @override
  Widget build(BuildContext context) {
    return Chart(
      data: const [
        {'category': 'A', 'group': 'Group 1', 'value': 10},
        {'category': 'A', 'group': 'Group 2', 'value': 15},
        {'category': 'B', 'group': 'Group 1', 'value': 8},
        {'category': 'B', 'group': 'Group 2', 'value': 12},
        {'category': 'C', 'group': 'Group 1', 'value': 20},
        {'category': 'C', 'group': 'Group 2', 'value': 5},
        // ... add more data for panning ...
        {'category': 'D', 'group': 'Group 1', 'value': 14},
        {'category': 'D', 'group': 'Group 2', 'value': 18},
        {'category': 'E', 'group': 'Group 1', 'value': 11},
        {'category': 'E', 'group': 'Group 2', 'value': 9},
      ],
      variables: {
        'category': Variable(
          accessor: (Map map) => map['category'] as String,
        ),
        'group': Variable(
          accessor: (Map map) => map['group'] as String,
        ),
        'value': Variable(
          accessor: (Map map) => map['value'] as num,
        ),
      },
      marks: [
        IntervalMark(
          position: Varset('category') * Varset('value') / Varset('group'),
          shape: ShapeEncode(value: RectShape()),
          // Use the state variable for the size ratio.
          size: SizeEncode(value: _barSizeRatio),
          modifiers: [
            // Use DodgeModifier for positioning, not resizing.
            DodgeModifier(ratio: 0.5),
          ],
        ),
      ],
      axes: [
        Defaults.horizontalAxis,
        Defaults.verticalAxis,
      ],
      coord: RectCoord(
        key: _coordKey,
        // Optional: Set an initial visible range if needed
        // initialRange: (0, 0.4),
      ),
      interactions: [
        // This interaction will update our bar size on scale.
        GestureInteraction(
          onScaleUpdate: (details) {
            // Get the current scale from the coordinate state.
            final coordState = _coordKey.currentState;
            if (coordState != null) {
              final scale = coordState.scale;
              setState(() {
                // Adjust the bar size based on the zoom scale.
                // You may need to tweak the multiplier for your desired effect.
                _barSizeRatio = 0.5 / scale;
              });
            }
          },
        ),
      ],
      // Enable selections and gestures for zooming and panning.
      selections: {
        'tap': PointSelection(
          on: {GestureType.scaleUpdate, GestureType.scroll},
          variable: 'category',
        ),
      },
      tooltip: TooltipGuide(
        followPointer: [false, true],
        align: Alignment.topLeft,
        offset: const Offset(-20, -20),
      ),
    );
  }
}
```

### Explanation of the Solution

1.  **State Management:** We introduce a state variable, `_barSizeRatio`, to hold the width ratio for our bars.
2.  **`CoordState` Key:** We assign a `GlobalKey<CoordState>` to our `RectCoord`. This gives us access to the coordinate system's state, including the current zoom `scale`.
3.  **`GestureInteraction`:** This is where the magic happens. We add a `GestureInteraction` and listen to the `onScaleUpdate` callback. This callback fires continuously during a pinch-to-zoom gesture.
4.  **Updating Bar Size:** Inside `onScaleUpdate`, we access the `_coordKey.currentState` to get the current `scale` of the horizontal axis. We then update our `_barSizeRatio` based on this scale. Dividing a base ratio by the scale ensures that as you zoom in (increasing `scale`), the bars get wider, and as you zoom out, they get narrower.
5.  **`SizeEncode`:** In the `IntervalMark`, we now use `SizeEncode(value: _barSizeRatio)`. This directly sets the width of the bars based on our state variable, bypassing the `DodgeSizeModifier`'s dynamic calculations.
6.  **`DodgeModifier`:** Notice we are still using `DodgeModifier`. This is important for positioning the bars side-by-side within each category. We've just removed its responsibility for sizing.

This approach gives you direct control over the bar sizing logic, tying it explicitly to the zoom gesture while ignoring panning. The result is that the bar widths will only change when you scale the graph, and they will remain consistent as you pan left and right.

I hope this detailed explanation and code example are helpful. Let me know if you have any more questions!
